pallets/free-calls/src/lib.rs

```diff
@@ -0,0 +1,430 @@
+        pub used_calls: NumberOfCalls,
+    }
+
+    impl<BlockNumber> ConsumerStats<BlockNumber> {
+        fn new(window_index: BlockNumber) -> Self {
+            ConsumerStats {
+                timeline_index: window_index,
+                used_calls: 0,
+            }
+        }
+    }
+
+    /// Configuration of a rate limiting window in terms of length and ratio to quota.
+    #[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug)]
+    pub struct WindowConfig<BlockNumber> {
+        /// The length of the window in number of blocks it will last.
+        pub period: BlockNumber,
+
+        /// The ratio between the quota and this window.
+        pub quota_ratio: QuotaToWindowRatio,
+    }
+
+    impl<BlockNumber> WindowConfig<BlockNumber> {
+        //TODO: try to also force period to be non zero.
+        pub const fn new(period: BlockNumber, quota_ratio: Option<QuotaToWindowRatio>) -> Self {
+            WindowConfig {
+                period,
+                quota_ratio: match quota_ratio {
+                    Some(non_zero) => non_zero,
+                    None => panic!("quota_ratio must be non zero"),
+                },
+            }
+        }
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub (super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_locker_mirror::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// The call type from the runtime which has all the calls available in your runtime.
+        type Call: Parameter
+            + Dispatchable<Origin = Self::Origin, PostInfo = PostDispatchInfo>
+            + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::Call>;
+
+        /// The configurations that will be used to limit the usage of the allocated quota to these
+        /// different configs.
+        #[pallet::constant]
+        type WindowsConfig: Get<Vec<WindowConfig<Self::BlockNumber>>>;
+
+        /// Filter on which calls are permitted to be free.
+        type CallFilter: Contains<<Self as Config>::Call>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+
+        /// A calculation strategy to convert locked tokens info to a quota.
+        type QuotaCalculationStrategy: QuotaCalculationStrategy<Self>;
+    }
+
+    /// Keeps track of each windows usage for each consumer.
+    #[pallet::storage]
+    #[pallet::getter(fn window_stats_by_consumer)]
+    pub(super) type WindowStatsByConsumer<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Twox64Concat,
+        // Index of the window in the list of window configurations.
+        WindowType,
+        ConsumerStats<T::BlockNumber>,
+    >;
```

Comment:
using a double map here does  not seem as efficient as a bounded vec.

from what i read, you are updating all the windows each time a tx is going though, so instead of modifying one storage item, you are modifying N storage items.